From 1f7af2ca88f116ea8171c7835cb72328b345b7ab Mon Sep 17 00:00:00 2001
From: "aperez@igalia.com"
 <aperez@igalia.com@268f45cc-cd09-0410-ab3c-d52691b4dbfc>
Date: Thu, 24 Aug 2017 08:45:07 +0000
Subject: [PATCH] Geoclue2 based backend should provide the right desktop ID
 https://bugs.webkit.org/show_bug.cgi?id=129879

Reviewed by Michael Catanzaro.

* platform/geoclue/GeolocationProviderGeoclue2.cpp:
(GeolocationProviderGeoclue::createGeoclueClientProxyCallback): Try first obtaining the application
identifier using GApplication first, keeping a fallback to the value returned by g_get_prgname().

git-svn-id: http://svn.webkit.org/repository/webkit/trunk@221132 268f45cc-cd09-0410-ab3c-d52691b4dbfc

diff --git a/Source/WebCore/platform/geoclue/GeolocationProviderGeoclue2.cpp b/Source/WebCore/platform/geoclue/GeolocationProviderGeoclue2.cpp
index 05bc672b525..4a16220240a 100644
--- a/Source/WebCore/platform/geoclue/GeolocationProviderGeoclue2.cpp
+++ b/Source/WebCore/platform/geoclue/GeolocationProviderGeoclue2.cpp
@@ -137,10 +137,16 @@ void GeolocationProviderGeoclue::createGeoclueClientProxyCallback(GObject*, GAsy
 
     // Geoclue2 requires the client to provide a desktop ID for security
     // reasons, which should identify the application requesting the location.
-    // FIXME: We provide the program name as the desktop ID for now but, in an ideal world,
-    // we should provide a proper "application ID" (normally the name of a .desktop file).
-    // https://bugs.webkit.org/show_bug.cgi?id=129879
-    geoclue_client_set_desktop_id(provider->m_clientProxy.get(), g_get_prgname());
+    // We use the application ID configured for the default GApplication, and
+    // also fallback to our old behavior of using g_get_prgname().
+    const char* appId = nullptr;
+    GApplication* defaultApp = g_application_get_default();
+    if (defaultApp)
+        appId = g_application_get_application_id(defaultApp);
+    if (!appId)
+        appId = g_get_prgname();
+
+    geoclue_client_set_desktop_id(provider->m_clientProxy.get(), appId);
 
     provider->startGeoclueClient();
 }
diff --git a/Source/WebKit2/UIProcess/API/gtk/WebKitGeolocationPermissionRequest.cpp b/Source/WebKit2/UIProcess/API/gtk/WebKitGeolocationPermissionRequest.cpp
index fb38df06042..a38e62d867e 100644
--- a/Source/WebKit2/UIProcess/API/gtk/WebKitGeolocationPermissionRequest.cpp
+++ b/Source/WebKit2/UIProcess/API/gtk/WebKitGeolocationPermissionRequest.cpp
@@ -39,6 +39,22 @@ using namespace WebKit;
  *
  * When a WebKitGeolocationPermissionRequest is not handled by the user,
  * it is denied by default.
+ *
+ * When embedding web views in your application, you *must* configure an
+ * application identifier to allow web content to use geolocation services.
+ * The identifier *must* match the name of the `.desktop` file which describes
+ * the application, sans the suffix.
+ *
+ * If your application uses #GApplication (or any subclass like
+ * #GtkApplication), WebKit will automatically use the identifier returned by
+ * g_application_get_application_id(). This is the recommended approach for
+ * enabling geolocation in applications.
+ *
+ * If an identifier cannot be obtained through #GApplication, the value
+ * returned by g_get_prgname() will be used instead as a fallback. For
+ * programs which cannot use #GApplication, calling g_set_prgname() early
+ * during initialization is needed when the name of the executable on disk
+ * does not match the name of a valid `.desktop` file.
  */
 
 static void webkit_permission_request_interface_init(WebKitPermissionRequestIface*);
-- 
2.13.5

